File: packages/core/src/worldid.ts
/**
 * World ID verification — Selfie Check (World ID 4.0 request flow).
 *
 * Selfie Check is a *medium*-assurance credential: liveness + facial similarity,
 * valid for a 90-day window, with no one-person-one-account guarantee. That makes
 * it exactly right for what we use it for — proving a live human is behind an
 * approval click at the moment money moves — and wrong for identity or Sybil
 * resistance. We never treat it as identity: the approver's *authority* comes
 * from the Privy key quorum, and Selfie Check only proves the click was live.
 *
 * Docs: https://docs.world.org/world-id/credentials/11
 *       https://docs.world.org/api-reference/developer-portal/verify
 */
import { env } from "./env";

/** The complete result IDKit hands back. Forward it verbatim — do not remap. */
export interface IDKitResult {
  protocol_version: string;
  nonce: string;
  responses: unknown[];
  user_presence_completed?: boolean;
}

export interface VerifyResult {
  ok: boolean;
  nullifier: string;
  environment: string;
  raw: Record<string, unknown>;
  error?: string;
  detail?: string;
}

/**
 * Verify server-side. Never call this from the browser.
 *
 * The v4 endpoint takes the whole IDKit result plus the action, and returns the
 * authoritative nullifier — we use *that*, never one the client sent us.
 */
export async function verifySelfieCheck(input: {
  result: IDKitResult;
  action: string;
}): Promise<VerifyResult> {
  const rpId = env.world.rpId();

  // Selfie Check access is gated, so keep the flow demoable without it.
  if (env.demo() || !rpId) {
    return {
      ok: true,
      nullifier: `demo_${input.result.nonce}`,
      environment: "demo",
      raw: { demo: true, action: input.action },
    };
  }

  const res = await fetch(`${env.world.base()}/api/v4/verify/${rpId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      ...input.result,
      action: input.action,
      environment: env.world.environment(),
    }),
  });

  const raw = (await res.json().catch(() => ({}))) as Record<string, unknown>;

  if (!res.ok || raw.success !== true) {
    return {
      ok: false,
      nullifier: "",
      environment: String(raw.environment ?? ""),
      raw,
      error: String(raw.code ?? res.status),
      detail: typeof raw.detail === "string" ? raw.detail : undefined,
    };
  }

  return {
    ok: true,
    nullifier: String(raw.nullifier ?? ""),
    environment: String(raw.environment ?? env.world.environment()),
    raw,
  };
}

/** Selfie Check credentials go stale after 90 days of inactivity. */
export const SELFIE_CHECK_TTL_MS = 90 * 24 * 60 * 60 * 1000;

/** How long a minted challenge stays valid. Matches IDKit's default RP TTL. */
export const CHALLENGE_TTL_SEC = 300;

/**
 * The signal binds the proof to one specific invoice + approver, so a proof
 * captured for invoice A can never be replayed to approve invoice B.
 */
export function approvalSignal(invoiceId: string, memberId: string): string {
  return `${invoiceId}:${memberId}`;
}

/** Human-readable copy for the error codes the verifier can return. */
export function explainVerifyError(code?: string, detail?: string): string {
  switch (code) {
    case "all_verifications_failed":
      return "The Selfie Check didn't pass. Try again in better lighting.";
    case "verification_error":
      return "That proof couldn't be verified. Please start a fresh check.";
    case "app_not_migrated":
      return "This app isn't on World ID 4.0 yet — check the Developer Portal.";
    case "user_presence_failed":
      return "Liveness check failed. Make sure it's you in front of the camera.";
    default:
      return detail ?? `Selfie Check failed${code ? ` (${code})` : ""}.`;
  }
}

File: server/internal/worldid/worldid.go
// Package worldid verifies World ID Selfie Check proofs.
//
// Selfie Check is a medium-assurance credential: liveness and facial
// similarity, valid 90 days, with no one-person-one-account guarantee. That
// makes it right for what we use it for — proving a live human was behind an
// approval click — and wrong for identity or Sybil resistance. Authority to
// approve comes from the Privy key quorum; this only proves the click was live.
package worldid

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// ChallengeTTL matches IDKit's default RP signature window.
const ChallengeTTL = 5 * time.Minute

// SelfieCheckValidity is how long a credential stays good without reuse.
const SelfieCheckValidity = 90 * 24 * time.Hour

// Result is the complete IDKit payload. Forward it verbatim — remapping
// response identifiers is explicitly called out as unsupported.
type Result struct {
	ProtocolVersion  string            `json:"protocol_version"`
	Nonce            string            `json:"nonce"`
	Responses        []json.RawMessage `json:"responses"`
	UserPresenceDone bool              `json:"user_presence_completed,omitempty"`
}

type Verification struct {
	OK          bool
	Nullifier   string
	Environment string
	Raw         json.RawMessage
	Code        string
	Detail      string
}

type Client struct {
	BaseURL     string
	RPID        string
	Environment string
	HTTP        *http.Client
	// Demo skips the network entirely so the flow stays demonstrable while
	// Selfie Check access is still gated.
	Demo bool
}

func New(baseURL, rpID, environment string, demo bool) *Client {
	return &Client{
		BaseURL:     baseURL,
		RPID:        rpID,
		Environment: environment,
		HTTP:        &http.Client{Timeout: 20 * time.Second},
		Demo:        demo || rpID == "",
	}
}

// Verify checks a proof server-side. Never call this from a browser.
//
// The nullifier we keep is the one the verifier returns — never one the client
// sent us, which would make the whole exercise decorative.
func (c *Client) Verify(ctx context.Context, result Result, action string) (Verification, error) {
	if c.Demo {
		return Verification{
			OK:          true,
			Nullifier:   "demo_" + result.Nonce,
			Environment: "demo",
			Raw:         json.RawMessage(`{"demo":true}`),
		}, nil
	}

	payload := map[string]any{
		"protocol_version":        result.ProtocolVersion,
		"nonce":                   result.Nonce,
		"responses":               result.Responses,
		"action":                  action,
		"environment":             c.Environment,
		"user_presence_completed": result.UserPresenceDone,
	}
	body, err := json.Marshal(payload)
	if err != nil {
		return Verification{}, err
	}

	url := fmt.Sprintf("%s/api/v4/verify/%s", c.BaseURL, c.RPID)
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return Verification{}, err
	}
	req.Header.Set("Content-Type", "application/json")

	res, err := c.HTTP.Do(req)
	if err != nil {
		return Verification{}, err
	}
	defer res.Body.Close()

	var parsed struct {
		Success     bool   `json:"success"`
		Nullifier   string `json:"nullifier"`
		Environment string `json:"environment"`
		Code        string `json:"code"`
		Detail      string `json:"detail"`
	}
	raw := new(bytes.Buffer)
	if _, err := raw.ReadFrom(res.Body); err != nil {
		return Verification{}, err
	}
	_ = json.Unmarshal(raw.Bytes(), &parsed)

	if res.StatusCode != http.StatusOK || !parsed.Success {
		return Verification{
			OK:     false,
			Raw:    raw.Bytes(),
			Code:   orDefault(parsed.Code, fmt.Sprintf("http_%d", res.StatusCode)),
			Detail: parsed.Detail,
		}, nil
	}

	return Verification{
		OK:          true,
		Nullifier:   parsed.Nullifier,
		Environment: orDefault(parsed.Environment, c.Environment),
		Raw:         raw.Bytes(),
	}, nil
}

// Signal binds a proof to one invoice and one approver, so a proof captured
// for one payout is meaningless against any other.
func Signal(invoiceID, memberID string) string {
	return invoiceID + ":" + memberID
}

// Explain turns verifier codes into copy a manager can act on.
func Explain(code, detail string) string {
	switch code {
	case "all_verifications_failed":
		return "The Selfie Check didn't pass. Try again in better lighting."
	case "verification_error":
		return "That proof couldn't be verified. Please start a fresh check."
	case "app_not_migrated":
		return "This app isn't on World ID 4.0 yet — check the Developer Portal."
	case "user_presence_failed":
		return "Liveness check failed. Make sure it's you in front of the camera."
	case "failed_by_host_app":
		// World App itself refused the request rather than producing a bad
		// proof — in practice the credential isn't available to that account:
		// Selfie Check not enabled for this app, or the wallet isn't enrolled.
		return "World App couldn't run the Selfie Check. The account needs to be " +
			"enrolled in Selfie Check, and the app needs it enabled in the Developer Portal."
	}
	if detail != "" {
		return detail
	}
	if code != "" {
		return "Selfie Check failed (" + code + ")."
	}
	return "Selfie Check failed."
}

func orDefault(v, fallback string) string {
	if v != "" {
		return v
	}
	return fallback
}
